File: src/engine/OpenGL/ShaderProgram.java
package engine.OpenGL;

import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.joml.Vector3f;
import org.joml.Vector4f;

import java.util.ArrayList;

import static org.lwjgl.opengl.GL20.*;

public class ShaderProgram {

	public static ArrayList<Integer> shaderProgramIDs = new ArrayList<>();

	private int id;

	public Shader[] shaders = new Shader[3];

	public static ShaderProgram currentShaderProgram;

	/**
	 * creates a new shader program
	 * @param vertPath path of the vertex shader
	 * @param fragPath path of the fragment shader
	 */
	public ShaderProgram(String vertPath, String fragPath) {
		id = glCreateProgram();
		shaderProgramIDs.add(id);

		shaders[0] = new Shader(vertPath, GL_VERTEX_SHADER);
		shaders[2] = new Shader(fragPath, GL_FRAGMENT_SHADER);

		glAttachShader(id, shaders[0].getID());
		glAttachShader(id, shaders[2].getID());

		glLinkProgram(id);

		if (glGetProgrami(id, GL_LINK_STATUS) != 1) {
			throw new RuntimeException("Failed to compile shader: ! " + glGetProgramInfoLog(id));
		}

		glDetachShader(id, shaders[0].getID());
		glDetachShader(id, shaders[2].getID());

		for (Shader s: shaders) {
			if (s != null) {
				for (Sampler sa : s.samplers) {
					sa.setLocation(id);
				}
			}
		}
	}

	/**
	 * creates a new shader program from a file name that includes a vertex and fragment shader
	 * @param folderName name of the folder in res/shaders
	 */
	public ShaderProgram(String folderName) {
		id = glCreateProgram();
		shaderProgramIDs.add(id);

		shaders[0] = new Shader("res/shaders/" + folderName + "/vert", GL_VERTEX_SHADER);
		shaders[2] = new Shader("res/shaders/" + folderName + "/frag", GL_FRAGMENT_SHADER);

		glAttachShader(id, shaders[0].getID());
		glAttachShader(id, shaders[2].getID());

		glLinkProgram(id);

		if (glGetProgrami(id, GL_LINK_STATUS) != 1) {
			throw new RuntimeException("Failed to compile shader: ! " + glGetProgramInfoLog(id));
		}

		glDetachShader(id, shaders[0].getID());
		glDetachShader(id, shaders[2].getID());

		for (Shader s: shaders) {
			if (s != null) {
				for (Sampler sa : s.samplers) {
					sa.setLocation(id);
				}
			}
		}
	}

	/**
	 * creates a new shader program
	 * @param shaderL list of the shaders, 0 represents the vertex shader, 1 represents the geometry shader, 2 represents the fragment shader
	 * @param exists marks if the corresponding shader exists or not, so if the vertex and fragment shader are used, but geometry shader is not, then the input should be [true, false, true]
	 */
	public ShaderProgram(Shader[] shaderL, boolean[] exists) {
		id = glCreateProgram();
		shaderProgramIDs.add(id);

		shaders = shaderL;

		if (exists[0]) {
			glAttachShader(id, shaders[0].getID());
		}
		if (exists[1]) {
			glAttachShader(id, shaders[1].getID());
		}
		if (exists[2]) {
			glAttachShader(id, shaders[2].getID());
		}

		glLinkProgram(id);

		if (exists[0]) {
			glDetachShader(id, shaders[0].getID());
		}
		if (exists[1]) {
			glDetachShader(id, shaders[1].getID());
		}
		if (exists[2]) {
			glDetachShader(id, shaders[2].getID());
		}

		for (Shader s: shaders) {
			if (s != null) {
				for (Sampler sa : s.samplers) {
					sa.setLocation(id);
				}
			}
		}
	}

	/**
	 * get the shader program handle
	 * @return shader program handle
	 */
	public int getID() {
		return id;
	}

	/**
	 * enables the shader program
	 */
	public void enable() {
		currentShaderProgram = this;
		glUseProgram(id);
		int textureNum = 0;
		for (Shader s: shaders) {
			if (s != null) {
				for (Sampler sa : s.samplers) {
					glUniform1i(sa.location, textureNum);
					++textureNum;
				}
			}
		}
	}

	/**
	 * binds the default shader
	 */
	public static void disable() {
		glUseProgram(0);
	}

	/**
	 * deletes the shader
	 */
	public void destroy() {
		glDeleteProgram(id);
		for (int i = 0; i < shaderProgramIDs.size(); ++i) {
			if (shaderProgramIDs.get(i) == id) {
				shaderProgramIDs.remove(i);
			}
		}
	}

	public void setUniform(int shader, int uniform, float val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, float vala, float valb) {
		shaders[shader].uniforms[uniform].set(vala, valb);
	}

	public void setUniform(int shader, int uniform, float vala, float valb, float valc) {
		shaders[shader].uniforms[uniform].set(vala, valb, valc);
	}

	public void setUniform(int shader, int uniform, float vala, float valb, float valc, float vald) {
		shaders[shader].uniforms[uniform].set(vala, valb, valc, vald);
	}

	public void setUniform(int shader, int uniform, Vector2f val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Vector3f val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Vector4f val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Matrix4f val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, int val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, float[] val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Vector2f[] val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Vector3f[] val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, Vector4f[] val) {
		shaders[shader].uniforms[uniform].set(val);
	}

	public void setUniform(int shader, int uniform, int[] val) {
		shaders[shader].uniforms[uniform].set(val);
	}
}

File: src/main/kotlin/engine/opengl/shaders/ShaderProgram.kt
package engine.opengl.shaders

import engine.opengl.GLResource
import engine.opengl.checkGLError
import java.lang.RuntimeException
import org.lwjgl.opengl.GL20.*
import java.util.ArrayList

import engine.opengl.shaders.ShaderType.*
import engine.resourceExists
import org.joml.*

private fun path(path : String) : String? = if (resourceExists(path)) path else null

@Suppress("EXPERIMENTAL_API_USAGE", "unused", "MemberVisibilityCanBePrivate")
class ShaderProgram : GLResource {
	val shaders : Array<Shader> = Array(3) {Shader.defaultShader}
	val uniforms : Array<Array<Uniform>>

	/**
	 * creates a new shader program
	 * @param vertPath path of the vertex shader
	 * @param fragPath path of the fragment shader
	 */
	constructor(vertPath: String?, geomPath : String?, fragPath: String?) : super(glCreateProgram()) {
		shaderProgramIDs.add(id)
		if (vertPath != null) {
			shaders[VERTEX_SHADER.index] = Shader(vertPath, VERTEX_SHADER)
			glAttachShader(id, shaders[VERTEX_SHADER.index].id)
		}
		if (geomPath != null) {
			shaders[GEOMETRY_SHADER.index] = Shader(geomPath, GEOMETRY_SHADER)
			glAttachShader(id, shaders[GEOMETRY_SHADER.index].id)
		}
		if (fragPath != null) {
			shaders[FRAGMENT_SHADER.index] = Shader(fragPath, FRAGMENT_SHADER)
			glAttachShader(id, shaders[FRAGMENT_SHADER.index].id)
		}
		glLinkProgram(id)
		if (glGetProgrami(id, GL_LINK_STATUS) != 1) {
			System.err.println("failed to compile shader")
			checkGLError()
			throw RuntimeException(glGetProgramInfoLog(id))//TODO create new exception for this
		}
		if (vertPath != null) glDetachShader(id, shaders[VERTEX_SHADER.index].id)
		if (geomPath != null) glDetachShader(id, shaders[GEOMETRY_SHADER.index].id)
		if (fragPath != null) glDetachShader(id, shaders[FRAGMENT_SHADER.index].id)
		uniforms = Array(shaders.size) {
			Array(shaders[it].uniformCount) {uniformIndex -> Uniform(shaders[it].uniforms[uniformIndex], id)}
		}
	}

	/**
	 * creates a new shader program from a file name that includes a vertex and fragment shader
	 * @param folderName name of the folder in res/shaders
	 */
	constructor(folderName: String) :
			this(path("res/shaders/$folderName/vert.glsl"),
				path("res/shaders/$folderName/geom.glsl"),
				path("res/shaders/$folderName/frag.glsl"))

	/**
	 * enables the shader program
	 */
	fun enable() {
		glUseProgram(id)
	}

	/**
	 * deletes the shader
	 */
	override fun destroy() {
		glDeleteProgram(id)
		for (i in shaderProgramIDs.indices) {
			if (shaderProgramIDs[i] == id) {
				shaderProgramIDs.removeAt(i)
				return
			}
		}
	}

	operator fun get(type : Int, index : Int) = uniforms[type][index]

	operator fun get(type : ShaderType, index : Int) = this[type.index, index]

	operator fun get(type : Int) = uniforms[type]

	operator fun get(type : ShaderType) = this[type.index]

	operator fun set(type : Int, index : Int, value : Float) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : FloatArray) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Double) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : DoubleArray) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Int) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : IntArray) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : UInt) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : UIntArray) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Boolean) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : BooleanArray) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector2fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector2fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector2dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector2dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector2ic) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector2ic>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector3fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector3fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector3dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector3dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector3ic) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector3ic>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector4fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector4fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector4dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector4dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Vector4ic) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Vector4ic>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix2fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix2fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix3fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix3fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix3x2fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix3x2fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix4fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix4fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix4x3fc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix4x3fc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix2dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix2dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix3dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix3dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix3x2dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix3x2dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix4dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix4dc>) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Matrix4x3dc) = this[type, index].set(value)
	operator fun set(type : Int, index : Int, value : Array<Matrix4x3dc>) = this[type, index].set(value)

	operator fun set(type : ShaderType, index : Int, value : Float) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : FloatArray) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Double) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : DoubleArray) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Int) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : IntArray) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : UInt) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : UIntArray) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Boolean) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : BooleanArray) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector2fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector2fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector2dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector2dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector2ic) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector2ic>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector3fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector3fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector3dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector3dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector3ic) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector3ic>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector4fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector4fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector4dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector4dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Vector4ic) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Vector4ic>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix2fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix2fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix3fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix3fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix3x2fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix3x2fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix4fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix4fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix4x3fc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix4x3fc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix2dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix2dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix3dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix3dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix3x2dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix3x2dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix4dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix4dc>) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Matrix4x3dc) = this[type, index].set(value)
	operator fun set(type : ShaderType, index : Int, value : Array<Matrix4x3dc>) = this[type, index].set(value)

	companion object {
		var shaderProgramIDs = ArrayList<Int>()

		/**
		 * binds the default shader
		 */
		fun disable() {
			glUseProgram(0)
		}
	}
}